### PR TITLE
bpo-31671: re.compile() cache checks flags type

### PR DESCRIPTION
**Pull request by vstinner**
Make sure that float flags are rejected even if the cache has an
entry with equal flags but flags of a different type.

https://bugs.python.org/issue31671

### CODE REVIEW

File: Lib/re.py [hunk 1]
```diff
@@ -274,8 +274,9 @@ def escape(pattern):
 _MAXCACHE = 512
 def _compile(pattern, flags):
     # internal: compile pattern
+    cache_key = (type(pattern), pattern, type(flags), flags)
```
Comment (vstinner) [inline Lib/re.py]: Putting type(flags) in the cache key prevents to merge flags=2 with flags=re.IGNORECASE.

I suggest to first focus on correctness. I let @methane handles the performance part with his PR #3862 :-)

But I propose to backport my fix to 2.7 and 3.6 without the optimization, so a minor performance regression for Python 3.6.

**Comment by kumbiepython** [comment]
Unsubscribe

On Oct 3, 2017 2:48 PM, "Victor Stinner" <notifications@github.com> wrote:

Make sure that float flags are rejected even if the cache has an
entry with equal flags but flags of a different type.
------------------------------
You can view, comment on, or merge this pull request online at:

  https://github.com/python/cpython/pull/3867
Commit Summary

   - bpo-31671: re.compile() cache checks flags type

File Changes

   - *M* Lib/re.py
   <https://github.com/python/cpython/pull/3867/files#diff-0> (5)
   - *M* Lib/test/test_re.py
   <https://github.com/python/cpython/pull/3867/files#diff-1> (17)
   - *A* Misc/NEWS.d/next/Library/2017-10-03-14-46-48.bpo-31671.5QpwsE.rst
   <https://github.com/python/cpython/pull/3867/files#diff-2> (3)

Patch Links:

   - https://github.com/python/cpython/pull/3867.patch
   - https://github.com/python/cpython/pull/3867.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/python/cpython/pull/3867>, or mute the thread

**Review by methane — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
I abandon my change for these reasons: https://bugs.python.org/issue31671#msg303753

**Pull request closed without merging**